File: go/grains/grains.go
package grains

import (
	"errors"
)

const chessSquares int = 64

func Square(x int) (grains uint64, err error) {
	if x < 1 || x > chessSquares {
		err = errors.New("number must be between 1-64")
	}
	return 1 << uint64(x-1), err
}

func Total() (total uint64) {
	for i := 1; i <= chessSquares; i++ {
		n, _ := Square(i)
		total += n
	}
	return total
}

File: solutions/ruby/grains/1/grains.rb
class Grains
  VALID_RANGE = (1..64)

  def self.square(number)
    raise ArgumentError unless VALID_RANGE.include? number

    2**(number - 1)
  end

  def self.total
    VALID_RANGE.map(&Grains.method(:square)).sum
  end
end
